src/deluge/gui/views/automation_instrument_clip_view.cpp

```diff
@@ -0,0 +1,3138 @@
+		}
+
+		bool isPatchCable = (modelStackWithParam->paramCollection
+		                     == modelStackWithParam->paramManager->getPatchCableSetAllowJibberish());
+		// Ok this is cursed, but will work fine so long as
+		// the possibly invalid memory here doesn't accidentally
+		// equal modelStack->paramCollection.
+
+		modelStackWithParam->autoParam->paste(startPos, endPos, scaleFactor, modelStackWithParam,
+		                                      &copiedParamAutomation, isPatchCable);
+
+		numericDriver.displayPopup(HAVE_OLED ? "Automation pasted" : "PASTE");
+		setDisplayParameterNameTimer();
+
+		if (playbackHandler.isEitherClockActive()) {
+			currentPlaybackMode->reversionDone(); // Re-gets automation and stuff
+		}
+
+		return;
+	}
+
+	numericDriver.displayPopup(HAVE_OLED ? "Can't paste automation" : "CANT");
+	setDisplayParameterNameTimer();
+}
+
+//select encoder action
+
+//used to change the parameter selection and reset shortcut pad settings so that new pad can be blinked
+//once parameter is selected
+void AutomationInstrumentClipView::selectEncoderAction(int8_t offset) {
+
+	//change midi CC or param ID
+	InstrumentClip* clip = getCurrentClip();
+	Instrument* instrument = (Instrument*)clip->output;
+
+	clip->lastSelectedParamShortcutX = kNoLastSelectedParamShortcutX;
+
+	if (instrument->type == InstrumentType::SYNTH || instrument->type == InstrumentType::KIT) {
+
+		if (!(instrument->type == InstrumentType::KIT && instrumentClipView.getAffectEntire())) {
+
+			if (clip->lastSelectedParamID == kNoLastSelectedParamID) {
+				clip->lastSelectedParamID = nonGlobalEffectableParamsForAutomation[0][1];
+				clip->lastSelectedParamType = nonGlobalEffectableParamsForAutomation[0][0];
+				clip->lastSelectedParamArrayPosition = 0;
+			}
+			else if ((clip->lastSelectedParamArrayPosition + offset) < 0) {
+				clip->lastSelectedParamID =
+				    nonGlobalEffectableParamsForAutomation[kNumNonGlobalEffectableParamsForAutomation - 1][1];
+				clip->lastSelectedParamType =
+				    nonGlobalEffectableParamsForAutomation[kNumNonGlobalEffectableParamsForAutomation - 1][0];
+				clip->lastSelectedParamArrayPosition = kNumNonGlobalEffectableParamsForAutomation - 1;
+			}
+			else if ((clip->lastSelectedParamArrayPosition + offset)
+			         > (kNumNonGlobalEffectableParamsForAutomation - 1)) {
+				clip->lastSelectedParamID = nonGlobalEffectableParamsForAutomation[0][1];
+				clip->lastSelectedParamType = nonGlobalEffectableParamsForAutomation[0][0];
+				clip->lastSelectedParamArrayPosition = 0;
+			}
+			else {
+				clip->lastSelectedParamID =
+				    nonGlobalEffectableParamsForAutomation[clip->lastSelectedParamArrayPosition + offset][1];
+				clip->lastSelectedParamType =
+				    nonGlobalEffectableParamsForAutomation[clip->lastSelectedParamArrayPosition + offset][0];
+				clip->lastSelectedParamArrayPosition += offset;
+			}
+		}
+
+		else if (instrument->type == InstrumentType::KIT && instrumentClipView.getAffectEntire()) {
```

Comment:
This condition will always be true, I think. So it would be good to check this in the `if` above, swap the bodies, and then just use `else`

Reply:
sorry can't find the original code this comment was associated with, I think I resolved it. Can you check? Thanks